src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -108,23 +131,107 @@ Score(const size_t queryIndex, TreeType& referenceNode)
 
     // Don't explore this tree branch.
     score = DBL_MAX;
+
+    // Store not used alpha for Monte Carlo.
+    if (kernelIsGaussian && monteCarlo)
+      accumMCAlpha(queryIndex) += depthAlpha;
+  }
+  else if (monteCarlo &&
+           referenceNode.NumDescendants() >= mcAccessCoef * initialSampleSize &&
+           newCalculations &&
+           kernelIsGaussian)
+  {
+    // Monte Carlo probabilistic estimation.
+    // Calculate z using accumulated alpha if possible.
+    const double alpha = depthAlpha + accumMCAlpha(queryIndex);
+    const boost::math::normal normalDist;
+    const double z =
+        std::abs(boost::math::quantile(normalDist, alpha / 2));
+
+    // Auxiliary variables.
+    const size_t numDesc = referenceNode.NumDescendants();
+    arma::vec sample;
+    size_t m = initialSampleSize;
+    double meanSample;
+    bool useMonteCarloPredictions = true;
+
+    // Resample as long as confidence is not high enough.
+    while (m > 0)
+    {
+      const size_t oldSize = sample.size();
+      const size_t newSize = oldSize + m;
+
+      // Don't use probabilistic estimation if this is going to take a similar
+      // amount of computations to the exact calculation.
+      if (newSize >= mcBreakCoef * numDesc)
+      {
+        useMonteCarloPredictions = false;
+        break;
+      }
+
+      // Increase the sample size.
+      sample.resize(newSize);
+      for (size_t i = 0; i < m; ++i)
+      {
+        // Sample and evaluate random points from the reference node.
+        const size_t randomPoint = math::RandInt(0, numDesc);
+        sample(oldSize + i) =
+            EvaluateKernel(queryIndex, referenceNode.Descendant(randomPoint));
+      }
+      meanSample = arma::mean(sample);
+      const double stddev = arma::stddev(sample);
+      const double mThreshBase =
+          z * stddev * (1 + relError) / (relError * meanSample);
+      const size_t mThresh = std::ceil(mThreshBase * mThreshBase);
+
+      if (sample.size() < mThresh)
+        m = mThresh - sample.size();
+      else
+        m = 0;
+    }
+
+    if (useMonteCarloPredictions)
+    {
+      // Confidence is high enough so we can use Monte Carlo estimation.
+      densities(queryIndex) += numDesc * meanSample;
+      score = DBL_MAX;
+
+      // Accumulated alpha has been used.
+      accumMCAlpha(queryIndex) = 0;
+    }
+    else
+    {
+      // Recurse.
+      score = minDistance;
+
+      if (referenceNode.IsLeaf() || tree::TreeTraits<TreeType>::HasSelfChildren)
+      {
+        // Reclaim not used alpha since the node will be exactly computed.
+        accumMCAlpha(queryIndex) += depthAlpha;
+      }
+    }
   }
   else
   {
     score = minDistance;
+
+    // If node is going to be exactly computed, reclaim not used alpha for
+    // Monte Carlo estimations.
+    if (kernelIsGaussian &&
+        monteCarlo &&
+        newCalculations &&
+        (referenceNode.IsLeaf() || tree::TreeTraits<TreeType>::HasSelfChildren))
```

Comment:
I think we should keep this line
```(referenceNode.IsLeaf() || tree::TreeTraits<TreeType>::HasSelfChildren)```.
`BaseCase()` will be called for every node that has not been pruned or already computed in CoverTree right?

Reply:
I'm thinking that the intent of this if statement is to only add to `accumMCAlpha` if it is not possible that Monte Carlo approximation will happen at any level deeper than this one.  Correct me if I'm wrong on that.

What you wrote about `BaseCase()` is basically right (if we change the word node to point), but just to restate, `BaseCase()` is called between the query point and every reference point in the reference node (i.e. not all the descendants, just the nodes held in the point, which for the cover tree is just one point, and for the kd-tree is zero points unless it's a leaf), and then recursion happens and `Score()` is called with any reference child nodes and the process repeats.

Anyway, I am not sure I'm answering the question very well; I think I've misunderstood what you mean.  Hopefully the answer is helpful, but if not, let me know what you're thinking and we can get it figured out. :)

Reply:
I think I now understand your point and it's right, cover tree is always a tricky one :)

So basically what this means is cover tree reclaims probability just when a node is pruned or when it gets to its leafs (and at that point it's useless since we're doing  a breadth-first traversal). Thanks for the clarification! :)

Reply:
Right, exactly. :+1: